rero_ils/modules/documents/utils_mef.py

```diff
@@ -0,0 +1,234 @@
+
+    mef_url = None
+    query = None
+
+    def __init__(self, run=False, verbose=False):
+        """Constructor."""
+        assert self.mef_url
+        assert self.query
+        self.count_found = {}
+        self.count_exists = {}
+        self.count_no_data = {}
+        self.count_no_mef = {}
+        self.verbose = verbose
+        if run:
+            self.process()
+
+    @abstractmethod
+    def get_local(self):
+        """Get local MEF record."""
+        raise NotImplementedError()
+
+    @abstractmethod
+    def get_online(self):
+        """Get online MEF record."""
+        raise NotImplementedError()
+
+    @abstractmethod
+    def process(self):
+        """Process replacement."""
+        raise NotImplementedError()
+
+    def increment_count(self, count, ref):
+        """Increment count.
+
+        :param count: count to use.
+        :param ref: reference to increment.
+        """
+        count.setdefault(ref, 0)
+        count[ref] += 1
+
+    @property
+    def counts(self):
+        """Counts."""
+        return (self.count_found, self.count_exists, self.count_no_data,
+                self.count_no_mef)
+
+    @property
+    def counts_len(self):
+        """Counts."""
+        return (len(self.count_found), len(self.count_exists),
+                len(self.count_no_data), len(self.count_no_mef))
+
+    def print_counts(self):
+        """Print counts."""
+        click.echo(
+            f'Found  : {len(self.count_found)} '
+            f'Exists : {len(self.count_exists)} '
+            f'No Data: {len(self.count_no_data)} '
+            f'No MEF : {len(self.count_no_mef)}'
+        )
+
+    def print_details(self):
+        """Print details count."""
+        if self.count_found:
+            click.echo('Found:')
+            for key in sorted(self.count_found.keys()):
+                click.echo(f'  {key:<20} : {self.count_found[key]}')
+        if self.count_exists:
+            click.echo('Exists:')
+            for key in sorted(self.count_exists.keys()):
+                click.echo(f'  {key:<20} : {self.count_exists[key]}')
+        if self.count_no_data:
+            click.echo('No Data:')
+            for key in sorted(self.count_no_data.keys()):
+                click.echo(f'  {key:<20} : {self.count_no_data[key]}')
+        if self.count_no_mef:
+            click.echo('No MEF:')
+            for key in sorted(self.count_no_mef.keys()):
+                click.echo(f'  {key:<20} : {self.count_no_mef[key]}')
+
+
+class ReplaceMefIdentifiedByContribution(ReplaceMefIdentifiedBy):
+    """Replace MEF identified by contribution class."""
+
+    def __init__(self, run=False, verbose=False):
+        """Constructor."""
+        self.mef_url = current_app.config.get('RERO_ILS_MEF_AGENTS_URL')
+        self.query = DocumentsSearch() \
+            .filter('exists', field='contribution.agent.identifiedBy')
+        self.cont_types = ['idref', 'gnd']
+        super().__init__(run=run, verbose=verbose)
```

Comment:
according to previous comment : 
```python
def __init__(self, run=False, verbose=False):
    mef_url = current_app.config.get('RERO_ILS_MEF_AGENTS_URL')
    super().__init__(mef_url, self._query_filter(), run=run, verbose=verbose)
    self.cont_types = ['idref', 'gnd']

def _query_filter():  # could be override by subclasses
     return DocumentsSearch() \
            .filter('exists', field='contribution.agent.identifiedBy')
```
It's in the best practice to call super as soon as possible. This is why I will set `cont_types` after `super` call.